{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Beginner quickstart to BioCypher (a practical example)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[//]: # (------------------------------------------    DO NOT MODIFY THIS    ------------------------------------------)\n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;\n",
    "      border-spacing:0;\n",
    "     }\n",
    ".tg td{border-color:black;\n",
    "       border-style:solid;\n",
    "       border-width:1px;\n",
    "       font-family:Arial, sans-serif;\n",
    "       font-size:14px;\n",
    "       overflow:hidden;\n",
    "       padding:10px 5px;\n",
    "       word-break:normal;\n",
    "      }\n",
    ".tg th{border-color:black;\n",
    "       border-style:solid;\n",
    "       border-width:1px;\n",
    "       font-family:Arial, sans-serif;\n",
    "       font-size:14px;\n",
    "       font-weight:normal;\n",
    "       overflow:hidden;\n",
    "       padding:10px 5px;\n",
    "       word-break:normal;\n",
    "      }\n",
    ".tg .tg-fymr{border-color:inherit;\n",
    "             font-weight:bold;\n",
    "             text-align:left;\n",
    "             vertical-align:top\n",
    "            }\n",
    ".tg .tg-0pky{border-color:inherit;\n",
    "             text-align:left;\n",
    "             vertical-align:top\n",
    "            }\n",
    "[//]: # (--------------------------------------------------------------------------------------------------------------)\n",
    "\n",
    "[//]: # (-------------------------------------    FILL THIS OUT WITH YOUR DATA    -------------------------------------)\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "    <tbody>\n",
    "      <tr>\n",
    "        <td class=\"tg-fymr\" style=\"font-weight: bold\">Title:</td>\n",
    "        <td class=\"tg-0pky\">Beginner quickstart to BioCypher (a practical example)</td>\n",
    "      </tr>\n",
    "      <tr>\n",
    "        <td class=\"tg-fymr\" style=\"font-weight: bold\">Authors:</td>\n",
    "        <td class=\"tg-0pky\">\n",
    "            <a href=\"https://github.com/ecarrenolozano\" target=\"_blank\" rel=\"noopener noreferrer\">Edwin Carreño</a>\n",
    "        </td>\n",
    "      </tr>\n",
    "      <tr>\n",
    "        <td class=\"tg-fymr\" style=\"font-weight: bold\">Affiliations:</td>\n",
    "        <td class=\"tg-0pky\">\n",
    "            <a href=\"https://www.ssc.uni-heidelberg.de/en\" target=\"_blank\" rel=\"noopener noreferrer\">Scientific Software Center</a>,\n",
    "            <a href=\"https://saezlab.org/\" target=\"_blank\" rel=\"noopener noreferrer\">Saez-Rodriguez Group</a>\n",
    "        </td>\n",
    "      </tr>\n",
    "      <tr>\n",
    "        <td class=\"tg-fymr\" style=\"font-weight: bold\">Date Created:</td>\n",
    "        <td class=\"tg-0pky\">19.03.2025</td>\n",
    "      </tr>\n",
    "      <tr>\n",
    "        <td class=\"tg-fymr\" style=\"font-weight: bold\">Description:</td>\n",
    "        <td class=\"tg-0pky\">A hands-on tutorial to explore BioCypher as quick as possible!</td>\n",
    "      </tr>\n",
    "    </tbody>\n",
    "</table>\n",
    "\n",
    "[//]: # (--------------------------------------------------------------------------------------------------------------)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "In this tutorial, we’re going to dive into **BioCypher** with a practical example. We’ll start with a simple CSV file and turn it into a Knowledge Graph step by step.\n",
    "\n",
    "At the end of this tutorial we will cover the following.\n",
    "    \n",
    "- Explore your data using Pandas\n",
    "\n",
    "- Graph modeling (from columns to nodes/edges)\n",
    "\n",
    "- BioCypher introduction:\n",
    "\n",
    "    - Setting up the configuration\n",
    "\n",
    "    - Writing a custom adapter to map your data to the graph model\n",
    "\n",
    "    - Importing everything into Neo4j to see a graph in action\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "\n",
    "Before starting our tutorial, ensure you have the following to guarantee a smooth experience!\n",
    "\n",
    "| Pre-requisite | Version   | Check on terminal |\n",
    "|---------------|-----------|-------------------|\n",
    "| Git           | >2.0      | `git --version`   |\n",
    "| Neo4j desktop | 2025.04.0 | `neo4j --version` |\n",
    "| Poetry        | >=1.8     | `poetry about`    |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clone repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'biocypher'...\n",
      "remote: Enumerating objects: 16138, done.\u001b[K\n",
      "remote: Counting objects: 100% (158/158), done.\u001b[K\n",
      "remote: Compressing objects: 100% (54/54), done.\u001b[K\n",
      "remote: Total 16138 (delta 65), reused 138 (delta 59), pack-reused 15980 (from 1)\u001b[K\n",
      "Receiving objects: 100% (16138/16138), 22.45 MiB | 40.90 MiB/s, done.\n",
      "Resolving deltas: 100% (11863/11863), done.\n"
     ]
    }
   ],
   "source": [
    "# Remove the biocypher folder if already exist\n",
    "%rm -rf biocypher\n",
    "!git clone https://github.com/biocypher/biocypher.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ecarreno/SSC-Projects/e-TUTORIALS/biocypher\n"
     ]
    }
   ],
   "source": [
    "%cd biocypher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating virtualenv \u001b[36mbiocypher\u001b[39m in /home/ecarreno/SSC-Projects/e-TUTORIALS/biocypher/.venv\n",
      "\u001b[2K\u001b[34mResolving dependencies...\u001b[39m \u001b[39;2m(0.6s)\u001b[39;22m\n",
      "\n",
      "\u001b[34mWriting lock file\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "!poetry lock --no-update\n",
    "!poetry install --no-root --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "### Importing Libraries\n",
    "\n",
    "Recommendations:\n",
    "\n",
    "- Respect the order of the imports, they are indicated by the numbers *1, 2, 3*.\n",
    "- One import per line is recommended, with this we can track easily any modified line when we use git.\n",
    "- Absolute imports are recommended (see *3. Local application/library specific imports* below), they improve readability and give better error messages.\n",
    "- You should put a blank line between each group of imports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Standard library imports\n",
    "import os\n",
    "\n",
    "# 2. Related third party imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# 3. Local application/library specific imports\n",
    "# import <mypackage>.<MyClass>         # this is an example\n",
    "# from <mypackage> import <MyClass>    # this is another example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 1. Exploratory Data Analysis with Pandas\n",
    "\n",
    "The CSV file we are going to use contains data about Protein-Protein interactions. It means in some way they should describe how the a Protein interacts with another Protein. For now, ignore what a Protein is and how they are interacting. Focus on how the information is structured.\n",
    "\n",
    "You should be able to answer the following:\n",
    "- How many columns does the dataset have?\n",
    "- How many rows?\n",
    "- All the columns are strings? numbers? other?\n",
    "\n",
    "**ONLY IF** you have experience in this kind of data, you can analyze further. \n",
    "- What types of interactions exists in this dataset?\n",
    "- Do the interactions contain properties?\n",
    "- A protein can contain properties? Which ones?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the path to the dataset (our CSV file)\n",
    "path_dataset = \"../subset_interactions_edgecases.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This file exist? True\n"
     ]
    }
   ],
   "source": [
    "print(\"This file exist? {}\".format(os.path.exists(path_dataset)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data into Pandas Dataframe (without predefined data types)\n",
    "\n",
    "By default the option *keep_default_na* is True, it means that Pandas will interpret empty values or null values as NaN values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframe = pd.read_table(path_dataset, sep=\"\\t\", keep_default_na=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ecarreno/SSC-Projects/e-TUTORIALS/biocypher\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>source_genesymbol</th>\n",
       "      <th>target_genesymbol</th>\n",
       "      <th>is_directed</th>\n",
       "      <th>is_stimulation</th>\n",
       "      <th>is_inhibition</th>\n",
       "      <th>consensus_direction</th>\n",
       "      <th>consensus_stimulation</th>\n",
       "      <th>consensus_inhibition</th>\n",
       "      <th>...</th>\n",
       "      <th>dorothea_coexp</th>\n",
       "      <th>dorothea_level</th>\n",
       "      <th>type</th>\n",
       "      <th>curation_effort</th>\n",
       "      <th>extra_attrs</th>\n",
       "      <th>evidences</th>\n",
       "      <th>ncbi_tax_id_source</th>\n",
       "      <th>entity_type_source</th>\n",
       "      <th>ncbi_tax_id_target</th>\n",
       "      <th>entity_type_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P1</td>\n",
       "      <td>P2</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>SMARCC1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P17947\",\"id_b\":\"Q9Y241\",\"positive\":[]...</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2</td>\n",
       "      <td>P1</td>\n",
       "      <td>SMARCC1</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P17947\",\"id_b\":\"Q9Y241\",\"positive\":[]...</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P2</td>\n",
       "      <td>P3</td>\n",
       "      <td>SMARCC1</td>\n",
       "      <td>Gtf2b</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"id_a\":\"Q92922\",\"id_b\":\"Q8NI51\",\"positive\":[]...</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "      <td>10116</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P3</td>\n",
       "      <td>P4</td>\n",
       "      <td>Gtf2b</td>\n",
       "      <td>Egr1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":false,\"DoRothEA_chipseq\":t...</td>\n",
       "      <td>{\"id_a\":\"P62916\",\"id_b\":\"P14882\",\"positive\":[]...</td>\n",
       "      <td>10116</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P4</td>\n",
       "      <td>P5</td>\n",
       "      <td>Egr1</td>\n",
       "      <td>Rheb</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P08046\",\"id_b\":\"P17563\",\"positive\":[]...</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P5</td>\n",
       "      <td>P6</td>\n",
       "      <td>Rheb</td>\n",
       "      <td>Myc</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>post_translational</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>{\"id_a\":\"Q921J2\",\"id_b\":\"Q6ZWX6\",\"positive\":[]...</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P6</td>\n",
       "      <td>P4</td>\n",
       "      <td>Myc</td>\n",
       "      <td>Egr1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P7</td>\n",
       "      <td>P8</td>\n",
       "      <td>CASP7</td>\n",
       "      <td>Cebpa</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P9</td>\n",
       "      <td>P9</td>\n",
       "      <td>ZNF318</td>\n",
       "      <td>ZNF318</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P1</td>\n",
       "      <td>P2</td>\n",
       "      <td>SPI1</td>\n",
       "      <td>SMARCC1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P17947\",\"id_b\":\"Q9Y241\",\"positive\":[]...</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P9</td>\n",
       "      <td>P4</td>\n",
       "      <td>Myc</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>D</td>\n",
       "      <td>transcriptional</td>\n",
       "      <td>0</td>\n",
       "      <td>{\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...</td>\n",
       "      <td>{\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "      <td>10090</td>\n",
       "      <td>protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   source target source_genesymbol target_genesymbol  is_directed  \\\n",
       "0      P1     P2              SPI1           SMARCC1            1   \n",
       "1      P2     P1           SMARCC1              SPI1            1   \n",
       "2      P2     P3           SMARCC1             Gtf2b            1   \n",
       "3      P3     P4             Gtf2b              Egr1            1   \n",
       "4      P4     P5              Egr1              Rheb            1   \n",
       "5      P5     P6              Rheb               Myc            1   \n",
       "6      P6     P4               Myc              Egr1            1   \n",
       "7      P7     P8             CASP7             Cebpa            1   \n",
       "8      P9     P9            ZNF318            ZNF318            1   \n",
       "9      P1     P2              SPI1           SMARCC1            1   \n",
       "10     P9     P4               Myc              EGR1            1   \n",
       "\n",
       "    is_stimulation  is_inhibition  consensus_direction  consensus_stimulation  \\\n",
       "0                0              0                    0                      0   \n",
       "1                0              0                    0                      0   \n",
       "2                0              0                    0                      0   \n",
       "3                0              0                    0                      0   \n",
       "4                0              0                    0                      0   \n",
       "5                0              0                    0                      0   \n",
       "6                0              0                    0                      0   \n",
       "7                0              0                    0                      0   \n",
       "8                0              0                    0                      0   \n",
       "9                0              0                    0                      0   \n",
       "10               0              0                    0                      0   \n",
       "\n",
       "    consensus_inhibition  ... dorothea_coexp dorothea_level  \\\n",
       "0                      0  ...          False              D   \n",
       "1                      0  ...          False              D   \n",
       "2                      0  ...            NaN            NaN   \n",
       "3                      0  ...          False              D   \n",
       "4                      0  ...          False              D   \n",
       "5                      0  ...            NaN            NaN   \n",
       "6                      0  ...          False              D   \n",
       "7                      0  ...          False              D   \n",
       "8                      0  ...          False              D   \n",
       "9                      0  ...          False              D   \n",
       "10                     0  ...          False              D   \n",
       "\n",
       "                  type  curation_effort  \\\n",
       "0      transcriptional                0   \n",
       "1      transcriptional                0   \n",
       "2      transcriptional                0   \n",
       "3      transcriptional                0   \n",
       "4      transcriptional                0   \n",
       "5   post_translational                1   \n",
       "6      transcriptional                0   \n",
       "7      transcriptional                0   \n",
       "8      transcriptional                0   \n",
       "9      transcriptional                0   \n",
       "10     transcriptional                0   \n",
       "\n",
       "                                          extra_attrs  \\\n",
       "0   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "1   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "2                                                  {}   \n",
       "3   {\"DoRothEA_curated\":false,\"DoRothEA_chipseq\":t...   \n",
       "4   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "5                                                  {}   \n",
       "6   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "7   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "8   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "9   {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "10  {\"DoRothEA_curated\":true,\"DoRothEA_chipseq\":fa...   \n",
       "\n",
       "                                            evidences  ncbi_tax_id_source  \\\n",
       "0   {\"id_a\":\"P17947\",\"id_b\":\"Q9Y241\",\"positive\":[]...                9606   \n",
       "1   {\"id_a\":\"P17947\",\"id_b\":\"Q9Y241\",\"positive\":[]...                9606   \n",
       "2   {\"id_a\":\"Q92922\",\"id_b\":\"Q8NI51\",\"positive\":[]...                9606   \n",
       "3   {\"id_a\":\"P62916\",\"id_b\":\"P14882\",\"positive\":[]...               10116   \n",
       "4   {\"id_a\":\"P08046\",\"id_b\":\"P17563\",\"positive\":[]...               10090   \n",
       "5   {\"id_a\":\"Q921J2\",\"id_b\":\"Q6ZWX6\",\"positive\":[]...               10090   \n",
       "6   {\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...               10090   \n",
       "7   {\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...               10090   \n",
       "8   {\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...               10090   \n",
       "9   {\"id_a\":\"P17947\",\"id_b\":\"Q9Y241\",\"positive\":[]...                9606   \n",
       "10  {\"id_a\":\"P18146\",\"id_b\":\"P09466\",\"positive\":[]...               10090   \n",
       "\n",
       "    entity_type_source  ncbi_tax_id_target  entity_type_target  \n",
       "0              protein                9606             protein  \n",
       "1              protein                9606             protein  \n",
       "2              protein               10116             protein  \n",
       "3              protein               10090             protein  \n",
       "4              protein               10090             protein  \n",
       "5              protein               10090             protein  \n",
       "6              protein               10090             protein  \n",
       "7              protein               10090             protein  \n",
       "8              protein               10090             protein  \n",
       "9              protein                9606             protein  \n",
       "10             protein               10090             protein  \n",
       "\n",
       "[11 rows x 36 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataframe.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 11 entries, 0 to 10\n",
      "Data columns (total 36 columns):\n",
      " #   Column                 Non-Null Count  Dtype \n",
      "---  ------                 --------------  ----- \n",
      " 0   source                 11 non-null     object\n",
      " 1   target                 11 non-null     object\n",
      " 2   source_genesymbol      11 non-null     object\n",
      " 3   target_genesymbol      11 non-null     object\n",
      " 4   is_directed            11 non-null     int64 \n",
      " 5   is_stimulation         11 non-null     int64 \n",
      " 6   is_inhibition          11 non-null     int64 \n",
      " 7   consensus_direction    11 non-null     int64 \n",
      " 8   consensus_stimulation  11 non-null     int64 \n",
      " 9   consensus_inhibition   11 non-null     int64 \n",
      " 10  sources                11 non-null     object\n",
      " 11  references             1 non-null      object\n",
      " 12  omnipath               11 non-null     bool  \n",
      " 13  kinaseextra            11 non-null     bool  \n",
      " 14  ligrecextra            11 non-null     bool  \n",
      " 15  pathwayextra           11 non-null     bool  \n",
      " 16  mirnatarget            11 non-null     bool  \n",
      " 17  dorothea               11 non-null     bool  \n",
      " 18  collectri              11 non-null     bool  \n",
      " 19  tf_target              11 non-null     bool  \n",
      " 20  lncrna_mrna            11 non-null     bool  \n",
      " 21  tf_mirna               11 non-null     bool  \n",
      " 22  small_molecule         11 non-null     bool  \n",
      " 23  dorothea_curated       9 non-null      object\n",
      " 24  dorothea_chipseq       9 non-null      object\n",
      " 25  dorothea_tfbs          9 non-null      object\n",
      " 26  dorothea_coexp         9 non-null      object\n",
      " 27  dorothea_level         9 non-null      object\n",
      " 28  type                   11 non-null     object\n",
      " 29  curation_effort        11 non-null     int64 \n",
      " 30  extra_attrs            11 non-null     object\n",
      " 31  evidences              11 non-null     object\n",
      " 32  ncbi_tax_id_source     11 non-null     int64 \n",
      " 33  entity_type_source     11 non-null     object\n",
      " 34  ncbi_tax_id_target     11 non-null     int64 \n",
      " 35  entity_type_target     11 non-null     object\n",
      "dtypes: bool(11), int64(9), object(16)\n",
      "memory usage: 2.4+ KB\n"
     ]
    }
   ],
   "source": [
    "dataframe.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset Metadata\n",
    "\n",
    "We are interested in having a table with the following information:\n",
    "\n",
    "\n",
    "- Column name.\n",
    "- Data Type.\n",
    "- A certain colum could contain null values.\n",
    "- Number of unique  values.\n",
    "\n",
    "That could be done using the next cell:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Unique values per column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column Name</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Nullable</th>\n",
       "      <th>Unique Values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>source</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>target</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>source_genesymbol</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>target_genesymbol</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>is_directed</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>is_stimulation</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>is_inhibition</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>consensus_direction</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>consensus_stimulation</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>consensus_inhibition</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sources</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>references</td>\n",
       "      <td>object</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>omnipath</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>kinaseextra</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ligrecextra</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pathwayextra</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>mirnatarget</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>dorothea</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>collectri</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tf_target</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>lncrna_mrna</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>tf_mirna</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>small_molecule</td>\n",
       "      <td>bool</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>dorothea_curated</td>\n",
       "      <td>object</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>dorothea_chipseq</td>\n",
       "      <td>object</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>dorothea_tfbs</td>\n",
       "      <td>object</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>dorothea_coexp</td>\n",
       "      <td>object</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>dorothea_level</td>\n",
       "      <td>object</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>type</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>curation_effort</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>extra_attrs</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>evidences</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ncbi_tax_id_source</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>entity_type_source</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ncbi_tax_id_target</td>\n",
       "      <td>int64</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>entity_type_target</td>\n",
       "      <td>object</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Column Name Data Type  Nullable  Unique Values\n",
       "0                  source    object     False              8\n",
       "1                  target    object     False              8\n",
       "2       source_genesymbol    object     False              8\n",
       "3       target_genesymbol    object     False              9\n",
       "4             is_directed     int64     False              1\n",
       "5          is_stimulation     int64     False              1\n",
       "6           is_inhibition     int64     False              1\n",
       "7     consensus_direction     int64     False              1\n",
       "8   consensus_stimulation     int64     False              1\n",
       "9    consensus_inhibition     int64     False              1\n",
       "10                sources    object     False              4\n",
       "11             references    object      True              1\n",
       "12               omnipath      bool     False              1\n",
       "13            kinaseextra      bool     False              2\n",
       "14            ligrecextra      bool     False              1\n",
       "15           pathwayextra      bool     False              1\n",
       "16            mirnatarget      bool     False              1\n",
       "17               dorothea      bool     False              2\n",
       "18              collectri      bool     False              1\n",
       "19              tf_target      bool     False              2\n",
       "20            lncrna_mrna      bool     False              1\n",
       "21               tf_mirna      bool     False              1\n",
       "22         small_molecule      bool     False              1\n",
       "23       dorothea_curated    object      True              2\n",
       "24       dorothea_chipseq    object      True              2\n",
       "25          dorothea_tfbs    object      True              1\n",
       "26         dorothea_coexp    object      True              1\n",
       "27         dorothea_level    object      True              1\n",
       "28                   type    object     False              2\n",
       "29        curation_effort     int64     False              2\n",
       "30            extra_attrs    object     False              3\n",
       "31              evidences    object     False              6\n",
       "32     ncbi_tax_id_source     int64     False              3\n",
       "33     entity_type_source    object     False              1\n",
       "34     ncbi_tax_id_target     int64     False              3\n",
       "35     entity_type_target    object     False              1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.DataFrame(\n",
    "    {\n",
    "        \"Column Name\": dataframe.columns,\n",
    "        \"Data Type\": dataframe.dtypes.values,\n",
    "        \"Nullable\": dataframe.isnull().any().values,\n",
    "        \"Unique Values\": [dataframe[col].nunique() for col in dataframe.columns],\n",
    "    }\n",
    ")\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To know all the unique values in a certain column, just type the column's name in the variable *field*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of unique values in field: type\n",
      "\t['post_translational', 'transcriptional']\n"
     ]
    }
   ],
   "source": [
    "field = \"type\"\n",
    "\n",
    "print(\n",
    "    \"List of unique values in field: {}\\n\\t{}\".format(\n",
    "        field, sorted(dataframe[field].unique())\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 2. Graph Modeling\n",
    "\n",
    "### What a graph is?\n",
    "\n",
    "A formal definition of a graph is: *\"A graph is a set V of vertices and a collection E of pairs of vertices from V, called edges.\"*\n",
    "\n",
    "$$\n",
    "G = (V,E)\n",
    "$$\n",
    "\n",
    "$\\lvert V \\rvert$ or $\\lvert n \\rvert$: number of vertices (nodes).\n",
    "$\\lvert E \\rvert$ or $\\lvert m \\rvert$: number of edges (relationships).\n",
    "\n",
    "For our purposes, a graph is just a collection of nodes interconnected between them by relationships. What a node is? What a relationship is? Instead of giving you formal definitions, let dive to this concept with two examples:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example 1. My friend Bob\n",
    "\n",
    "```mermaid\n",
    "graph LR\n",
    "    Alice((Alice)) -- is a friend of --> Bob((Bob))\n",
    "    Bob((Bob)) -- is a --> Person((Person))\n",
    "    Bob((Bob)) -- is interested in --> IC((Integrated Circuit))\n",
    "    IC((Integrated Circuit)) -- was invented by --> JK((Jack Kilby))\n",
    "    IC((Integrated Circuit)) -- was invented by --> RN((Robert Noyce))\n",
    "    JK((Jack Kilby)) -- worked for --> TI((Texas Instruments))\n",
    "    JK((Jack Kilby)) -- is a --> Person((Person))\n",
    "    RN((Robert Noyce)) -- worked for --> FS((Fairchild Semiconductors))\n",
    "    RN((Robert Noyce))-- is a --> Person((Person))\n",
    "    TI((Texas Instruments)) -- is a --> Company((Company))\n",
    "    FS((Fairchild Semiconductors)) -- is a --> Company((Company))\n",
    "    \n",
    "```\n",
    "\n",
    "##### List of Nodes\n",
    "\n",
    "They represent abstract things, concepts, and usually can be identified by **nouns**.\n",
    "\n",
    "- Alice\n",
    "- Bob\n",
    "- Person\n",
    "- Integrated circuit\n",
    "- Jack Kilby\n",
    "- Robert Noyce\n",
    "- Texas Instruments\n",
    "- Fairchild Semiconductors\n",
    "- Company\n",
    "\n",
    "##### List of Relationships (Edges)\n",
    "\n",
    "They represent actions, interactions, and usually can be identified by **verbs**.\n",
    "\n",
    "- *is a friend of*\n",
    "- *is a*\n",
    "- *is interested in*\n",
    "- *was invented by*\n",
    "- *worked for*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example 2. Biological Organization Underlying Human Disease\n",
    "\n",
    "```mermaid\n",
    "graph LR\n",
    "    Tissue((Tissue)) -- contains --> DNA((DNA))\n",
    "    DNA((DNA)) -- encodes --> Gene((Gene))\n",
    "    Gene((Gene)) -- interact_with --> Gene((Gene))\n",
    "    Variant((Variant)) -- alters --> Gene((Gene))\n",
    "    Variant((Variant)) -- causes_or_contributes_to --> Disease((Disease))\n",
    "    Gene((Gene)) -- causes_or_contributes_to --> Disease((Disease))\n",
    "    DNA((DNA)) -- transcribed_into --> RNA((RNA))\n",
    "    RNA((RNA)) -- processed_into --> miRNA((miRNA))\n",
    "    miRNA((RNA)) -- decreases --> mRNA((mRNA))\n",
    "    RNA((RNA)) -- processed_into --> mRNA((RNA))\n",
    "    mRNA((mRNA)) -- translated_into --> Protein((Protein))\n",
    "    Protein((Protein)) -- has_function --> MolFunc((Molecular Function))\n",
    "    Protein((Protein)) -- participates_in --> BioProc((Biological Process))\n",
    "    Protein((Protein)) -- located_in --> CellComp((Cellular Component))\n",
    "    Pathway((Pathway)) -- has_component --> CellComp((Cellular Component))\n",
    "    Pathway((Pathway)) -- realizes --> MolFunc((Molecular Function))\n",
    "    Pathway((Pathway)) -- has_part --> BioProc((Biological Process))\n",
    "    Protein((Protein)) -- participates_in --> Pathway((Pathway))\n",
    "    Chemical((Chemical)) -- participates_in --> Pathway((Pathway))\n",
    "    Chemical((Chemical)) -- is_substance_that_treats --> Disease((Disease))\n",
    "    Pathway((Pathway)) -- causes_or_contributes_to --> Disease((Disease))\n",
    "```\n",
    "\n",
    "##### List of Nodes\n",
    "\n",
    "They represent abstract things, concepts, and usually can be identified by **nouns**.\n",
    "\n",
    "- Biological Process\n",
    "- Cell\n",
    "- Chemical\n",
    "- Disease\n",
    "- DNA\n",
    "- Fluid\n",
    "- Gene\n",
    "- miRNA\n",
    "- Molecular Function\n",
    "- mRNA\n",
    "- Pathway\n",
    "- Protein\n",
    "- RNA\n",
    "- Tissue\n",
    "- Variant\n",
    "\n",
    "##### List of Relationships (Edges)\n",
    "\n",
    "They represent actions, interactions, and usually can be identified by **verbs**.\n",
    "\n",
    "- *alters*\n",
    "- *causes_or_contributes_to*\n",
    "- *contains*\n",
    "- *decreases*\n",
    "- *encodes*\n",
    "- *has_component*\n",
    "- *has_function*\n",
    "- *has_part*\n",
    "- *interacts_with*\n",
    "- *is_substance_that_treats*\n",
    "- *located_in*\n",
    "- *participates_in*\n",
    "- *processed_into*\n",
    "- *realizes*\n",
    "- *transcribed_into*\n",
    "- *translated_into*\n",
    "\n",
    "**Reference:** inspired on [An open source knowledge graph ecosystem for the life sciences (2024)](https://www.nature.com/articles/s41597-024-03171-w)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Key points\n",
    "\n",
    "1. A knowledge graph is a collection of nodes and edges that define interactions between them.\n",
    "2. Nodes and Edges contains properties i.e., A person, has a name, an ID, a date of birth, etc.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Modelling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do not worry if the Proteins and interactions are not in your scope of study. Just ignore that, and treat those as nodes and edges.\n",
    "\n",
    "#### First iteration:\n",
    "Our input is the data contained in the dataset, this data contains information about Protein-Protein interactions. As we saw, nodes and edges can contain properties associated. So, the main challenge here is to create a diagram similar to the examples we have. Do not worry, we already have two keys to achieve this:\n",
    "\n",
    "1. The *\"Protein-protein interactions\"* phrase.\n",
    "2. Our dataset has two columns called \"source\" and \"target\".\n",
    "\n",
    "It means we have a Protein node and another Protein node. At the same time, a relatioship between them exist. That is all you need to know right now.\n",
    "\n",
    "```mermaid\n",
    "graph LR\n",
    "    Source((Protein)) -- interacts_with --> Target((Protein))\n",
    "```\n",
    "#### Second iteration\n",
    "Can we improve this graph?- the answer is yes, take a look again at the data. Exist a field called \"type\". This column contains two possible values:  \n",
    "\n",
    "- *post_translational*\n",
    "- *transcriptional*\n",
    "\n",
    "We can change the edge for a more descriptive one, such as *is_interaction_type*.\n",
    "\n",
    "```mermaid\n",
    "graph LR\n",
    "    A((Protein))\n",
    "    B((Protein))\n",
    "    A -- is_interaction_type --> B\n",
    "```\n",
    "\n",
    "#### Third iteration\n",
    "In this point, Nodes and Edges do not contain properties. Can we add properties to the Nodes and Edges? Yes, we can. Analyze the dataset again, for this tutorial I have choosen the following properties to assign to each node.\n",
    "\n",
    "**Nodes**\n",
    "- Protein node (source):\n",
    "    - source_genesymbol\n",
    "    - source_ncbi_tax_id\n",
    "    - source_entity_type\n",
    "\n",
    "- Protein node (target):\n",
    "    - target_genesymbol\n",
    "    - target_ncbi_tax_id\n",
    "    - target_entity_type\n",
    "\n",
    "**Edges**\n",
    "- *is_interaction_type*:\n",
    "    - is_directed\n",
    "    - is_stimulation\n",
    "    - is_inhibition\n",
    "    - sources\n",
    "\n",
    "\n",
    "<div>\n",
    "<img src=\"./protein-protein-interaction-model2.png\" width=\"300\"/>\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manual checking\n",
    "\n",
    "By looking at the data (source and target columns) try to draw nodes and edges in a piece of paper. At the end your data should represent a graph like this one:\n",
    "\n",
    "<div>\n",
    "<img src=\"./protein-protein-interaction-data.png\" width=\"400\"/>\n",
    "</div>\n",
    "\n",
    "A graph like this one is what we want, by using BioCypher we can create it and display it in Neo4j.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 3. BioCypher Adapter\n",
    "\n",
    "In order to build a graph by using BioCypher we need to do the following:\n",
    "\n",
    "1. Read the data (our CSV file) and generate Nodes and Edges (Adapter concept)\n",
    "2. Define the schema for the graph (which types of nodes/edges and their properties should we expect in our graph)\n",
    "3. Configure BioCypher to write an script that will serve to import the BioCypher graph into a Neo4j graph.\n",
    "\n",
    "That's all, we will explain each step in this tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
